zk.el

```diff
@@ -495,46 +495,19 @@ supplied. Can take a PROMPT argument."
          ,file)))
    (zk--directory-files t)))
 
-(defun zk--parse-id (target ids &optional zk-alist)
-  "Return TARGET, either `file-path or `title, from files with IDS.
-Takes a single ID, as a string, or a list of IDs. Takes an
-optional ZK-ALIST, for efficiency if `zk--parse-id' is called
-in an internal loop."
-  (cond
-   ((and (eq target 'file-path)
-         (stringp ids))
-    (car (zk--directory-files t ids)))
-   ((and (eq target 'file-path)
-         (zk--singleton-p ids))
-    (car (zk--directory-files t (car ids))))
-   (t
-    (let* ((zk-alist (or zk-alist
-                         (zk--alist)))
-           (zk-id-list (zk--id-list))
-           (return
-            (cond ((eq target 'file-path)
-                   (cond ((stringp ids)
-                          (if (member ids zk-id-list)
-                              (cddr (assoc ids zk-alist))
-                            (user-error "No file associated with %s" ids)))
-                         ((listp ids)
-                          (mapcar
-                           (lambda (x)
-                             (caddr (assoc x zk-alist)))
-                           ids))))
-                  ((eq target 'title)
-                   (cond ((stringp ids)
-                          (if (member ids zk-id-list)
-                              (cadr (assoc ids zk-alist))
-                            (user-error "No file associated with %s" ids)))
-                         ((listp ids)
-                          (mapcar
-                           (lambda (x)
-                             (cadr (assoc x zk-alist)))
-                           ids)))))))
-      (if (zk--singleton-p return)
-          (car return)
-        return)))))
+(defun zk--parse-id (target id &optional zk-alist)
+  "Return TARGET, either `file-path or `title, from file with ID.
+Takes a single ID, as a string. Takes an optional ZK-ALIST, for
+efficiency if `zk--parse-id' is called in an internal loop."
+  (let* ((zk-alist (or zk-alist (zk--alist)))
+         (zk-id-list (zk--id-list nil zk-alist)))
+    (unless (member id zk-id-list)
+      (user-error "No file associated with %s" id))
+    (cond ((eq target 'file-path)
+           (caddr (assoc id zk-alist)))
+          ((eq target 'title)
+           (cadr (assoc id zk-alist)))
+          (t (error "Invalid target: %s" target)))))
```

Comment:
Running `zk--alist` to get the file-path for a single id is more resource intensive than running `zk--directory-files`:

(benchmark-run 10
  (rb/zk--parse-id 'file-path "202303261345"))
(5.010084 20 3.469380000000001)

(benchmark-run 10
  (zk--parse-id 'file-path "202303261345"))
(1.939562 10 1.7060150000000078)

Reply:
Ah, good eye. After the added commits to this branch:

```
(benchmark-run 100 (zk--parse-id 'file-path "202206052002"))
(10.382485999999998 100 9.959769999999999)

(benchmark-run 100 (rb/zk--parse-id 'file-path "202206052002"))
(0.35399400000000003 0 0.0)
```

Reply:
That last benchmark number can't be right. Maybe you can double check, because here is what I'm getting on this branch.

(benchmark-run 10
  (zk--parse-id 'file-path "201804151211"))
(15.371123 0 0.0)
(10.826334 0 0.0)
(8.144576 0 0.0)
(5.173069 0 0.0)
(2.360358 0 0.0)
(0.740523 0 0.0)

These are the results of benchmark runs, in reverse order. What it shows is that each successive run is slower and slower. This is because your code forgoes the forced `garbage-collect` that is built in to `zk--directory-files`.

All this aside, I'm not exactly sure what concrete problem this PR is trying to address. The reasons listed above are abstract --- simplifying, error checking, and consistency. I'm not sure these warrant significant rewriting/restructuring of the code, unless there is some major improvement to be had (not just a few milliseconds here and there).

Reply:
The benchmark is right, but it requires the two commits (2999c7d5290e3b83409ce1bcc9b2acee4593dc83 030df971c4e74bee118a4978e5c76a90e93b1ebf) I added below. There is no garbage to collect, since `zk--id-file` (which is then used to rewrite `zk--parse-id`) simply looks for the ID in the file system rather than getting a list of all zk-files in `zk-directory` and then going through them to find the ID.

Reply:
> All this aside, I'm not exactly sure what concrete problem this PR is trying to address. The reasons listed above are abstract --- simplifying, error checking, and consistency. I'm not sure these warrant significant rewriting/restructuring of the code, unless there is some major improvement to be had (not just a few milliseconds here and there).

I appreciate your feedback and understand your concerns regarding the concrete benefits of the changes. However, improvements in code design and refactoring are not just about performance boosts, but also about maintainability and scalability. When code is simplified and consistent, it becomes easier for contributors to understand and work with it. This, in turn, leads to fewer bugs and makes it easier to add new features in the future.

I understand that the reasons I provided might seem abstract, but they are fundamental principles in software engineering. They help in ensuring that as the project grows, it remains manageable and doesn’t become too complex to handle. A lot of what I'm doing has been influenced by a well-regarded software engineering book by Martin Fowler /Refactoring: Improving the Design of Existing Code/ (1999, 2019). Attached is an excerpt about the principles of refactoring, which also goes into the whys of refactoring.

In particular, the most useful parts:
- Pages 50-51: Discusses the Design Stamina Hypothesis, which explains how good internal design increases the stamina of the software effort, allowing for faster and more sustainable development. It also talks about when refactoring should be done and how it fits into the workflow.
- Pages 54-55: Explains how to communicate the importance of refactoring to managers and stakeholders who may not have technical awareness. It also discusses when not to refactor and the problems that can arise with refactoring.
- Pages 62-63: How refactoring affects software architecture and how it can be used to form a well-designed code base that can adapt to changing needs. It emphasizes the importance of building software that is excellently designed for current needs and using refactoring to adapt as requirements change.
- Page 67: Provides a historical background on where refactoring came from and how it has been an essential practice among good programmers for maintaining clean code.

I only have a sporadic education in software engineering (was fascinated with programming as a teenager, began as a CS major in college, but then moved to humanities), but I have been programming in some capacity since the 1990s. I read Fowler's book shortly after finding your `zk` project, actually, which helped to explain the reasons behind many practices that I've picked up over the years from coding style guides, programming language manuals, and other software development books I've perused. 

I imagine you're quite busy as an assistant professor, but if you ever get a chance, Fowler's book is excellent, though quite a few refactorings aren't very applicable to Emacs Lisp.

[Fowler2019, Excerpt on Principles in Refactoring.pdf](https://github.com/localauthor/zk/files/11923435/Fowler2019.Excerpt.on.Principles.in.Refactoring.pdf)

Reply:
P.S. After re-reading my previous comment, I'd like to clarify a couple of things:

1. I really appreciate the way you designed and implemented `zk`, `zk-index`, and `zk-desktop`. As soon as I found them, I jumped head-first, decoupling my own half-baked Zettelkasten-like implementation from `deft` that I was using until then (and finding difficult to work with) and working to add features to `zk` that I found useful in my own work but that were missing. This was the first time I contributed to a github project, so I'm also grateful to your patience as I continue to learn about the process.

2. The bit about my programming experience was meant to both emphasize that I am in no way a professional, so there are certainly many areas and topics that I am ignorant about. Collaborating with others is definitely one of them, since I've never worked in software development or participated in team-driven programming.

3. I don't remember if I mentioned it, but as a late-stage Ph.D. student in environmental humanities who is still struggling with the dissertating part, I am excited to find another Zettelkasten-and-Emacs aficionado. Especially since we're more or less in the same field (I suspect it has something to do with ecology being tightly linked with systems thinking). In other words, I'm still only aspiring to be an assistant professor. Perhaps if I spent less time messing with Emacs and more time writing, that would speed things up.